### PR TITLE
fix(parsing): normalize nested column names

### PR DESCRIPTION
**Pull request by shcheklein**
Part of the https://github.com/iterative/datachain/issues/481

Fixes a few issue with nested column names. We had before a [function](https://github.com/iterative/datachain/blob/main/src/datachain/lib/arrow.py#L146-L156) that was normalizing the top level schema names in a single place [here](https://github.com/iterative/datachain/blob/main/src/datachain/lib/arrow.py#L131).

It means that if you have a JSON file to parse with `parse_tabular` (yes, it supports JSONs as well), or any other nested structure where we create Pydantic models for nested fields with [`dict_to_data_model`](https://github.com/iterative/datachain/blob/main/src/datachain/lib/data_model.py#L62-L68) we either had a bunch of non-normalized names except for the top level, or we could end up with a syntax error.

E.g. I had a JSON like this:

```json
{
  "bff_contained_ngram_count_before_dedupe": 0,
  "language_id_whole_page_fasttext": {
    "en": 0.9512948989868164
  },
  "metadata": {
    "Content-Length": "112946",
    "Content-Type": "application/http; msgtype=response",
    "WARC-Block-Digest": "sha1:U462KE2IDQWYDJQDRXIP3UVH465G2I2G",
    "WARC-Concurrent-To": "<urn:uuid:eaf4d817-f140-4a5e-9fa9-64d0dcb32a08>",
    "WARC-Date": "2017-05-27T21:20:09Z",
    "WARC-IP-Address": "198.57.149.47",
    "WARC-Identified-Payload-Type": "text/html",
    "WARC-Payload-Digest": "sha1:KAFZ77S3HPJ7SOABQ4ZMMVT47OCXZR6R",
    "WARC-Record-ID": "<urn:uuid:f8557905-b0dd-47b9-bc60-3106b3c18b4c>",
    "WARC-Target-URI": "http://www.muslimlinkpaper.com/index.php/editors-desk/13-letter-to-the-editor/3054-a-letter-to-president-obama.html",
    "WARC-Truncated": "length",
    "WARC-Type": "response",
    "WARC-Warcinfo-ID": "<urn:uuid:398fb229-e8ed-4f8b-86af-00eb8d6b2594>"
  },
  "previous_word_count": 178,
  "text": "A Letter ...",
  "url": "http://www.muslimlinkpaper.com/index.php/editors-desk/13-letter-to-the-editor/3054-a-letter-to-president-obama.html",
  "warcinfo": "robots: classic\r\nhostname: ip-10-185-224-210.ec2.internal\r\nsoftware: Nutch 1.6 (CC)/CC WarcExport 1.0\r\nisPartOf: CC-MAIN-2017-22\r\noperator: Common Crawl Admin\r\ndescription: Wide crawl of the web for May 2017\r\npublisher: Common Crawl\r\nformat: WARC File Format 1.0\r\nconformsTo: http://bibnum.bnf.fr/WARC/WARC_ISO_28500_version1_latestdraft.pdf",
  "fasttext_openhermes_reddit_eli5_vs_rw_v2_bigram_200k_train_prob": 0.967424213886261
}
```

Particularly, dealing with a field `WARC-Concurrent-To` is problematic. `sqlalchmey` fails to quote bound params in `INSERT ... VALUES ...` and `-TO` triggers a syntax error, as well as `-SELECT` for example.

With this PR, we are applying the same rules to all column names - top level and nested. It requires making an alias on Pydantic model, but I hope that is fine.

## TODO:

- [x] Add more unit tests for the new name normalization function
- [x] To @0x2b3bfa0 point - make it even more strict (make them valid Python identifiers - e.g. they can't start with numbers) - at least in case when we generate a pydantic model out of them.

### CODE REVIEW

File: src/datachain/lib/utils.py [hunk 1]
```diff
@@ -28,3 +30,19 @@ def __init__(self, message):
...
         super().__init__(f"Error for column {col_name}: {msg}")
+
+
+def normalize_col_names(col_names: Sequence[str]) -> dict[str, str]:
+    default_column = 0
+    converted_col_names = {}
+    for column in col_names:
+        original_column = column
+        column = column.lower()
+        column = re.sub("-+", "_", column)
```
Comment (shcheklein) [inline src/datachain/lib/utils.py:41]: C: I've changed the rule a bit here as well. It now changes `-` into `_` and multiple `_` into a single one. 

Also checks if the column name after normalization exists it now also applies the `default_column` generator.

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py:41]: @shcheklein, quick question: what happens if the original column name _begins_ with a number, would it break something?

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py:41]: Asking just in case we use these identifiers as e.g. Python attributes, where `column1` would be valid but `1column` wouldn't

Comment (shcheklein) [reply src/datachain/lib/utils.py]: good question, I'm a bit hesitant to go that far (not always we have to create a pydantic model - e.g. we might not need for top level fields - just a flat parquet / CSV). I feel it's a bit broader topic ... I'll think about it and if we need to solve it in this PR

Comment (shcheklein) [reply src/datachain/lib/utils.py]: probably at least in the `dict_to_data_model` we have to ensure this I guess, so can introduce an additional param for that

File: src/datachain/lib/utils.py [hunk 2]
```diff
@@ -28,3 +30,29 @@ def __init__(self, message):
...
+def normalize_col_names(col_names: Sequence[str]) -> dict[str, str]:
+    gen_col_counter = 0
+    new_col_names = {}
+    org_col_names = set(col_names)
+
+    for org_column in col_names:
+        new_column = org_column.lower()
+        new_column = re.sub("[-_\\s]+", "_", new_column)
+        new_column = new_column.strip("_")
+        new_column = re.sub("[^0-9a-z_]+", "", new_column)
```
Comment (0x2b3bfa0) [inline src/datachain/lib/utils.py:44]:
```suggestion
        new_column = org_column.lower()
        new_column = re.sub("[^0-9a-z]+", "_", new_column)
        new_column = new_column.strip("_")
```

Wouldn't you prefer this alternative normalization, which interprets any non-alphanumeric character as a separator?
* Before: `Old/Man's@Haven 1` → `oldmanshaven_1`
* After: `Old/Man's@Haven@ 1` → `old_man_s_haven_1`

File: src/datachain/lib/utils.py [hunk 3]
```diff
@@ -28,3 +30,29 @@ def __init__(self, message):
...
+
+        if not new_column or (new_column != org_column and new_column in org_col_names):
+            while True:
+                generated_column = f"c{gen_col_counter}"
+                gen_col_counter += 1
+                if new_column:
+                    generated_column = f"{generated_column}_{new_column}"
+                if generated_column not in org_col_names:
+                    new_column = generated_column
+                    break
```
Comment (0x2b3bfa0) [inline src/datachain/lib/utils.py:54]:
```suggestion
        if not new_column or (new_column != org_column and new_column in new_col_names):
            while True:
                generated_column = f"c{gen_col_counter}"
                gen_col_counter += 1
                if new_column:
                    generated_column = f"{generated_column}_{new_column}"
                if generated_column not in new_col_names:
                    new_column = generated_column
                    break
```

I thought the whole point of this logic was checking for duplicates in the `new_column_names` dictionary?

### Test vector

```python
normalize_col_names(("column+name", "column=name"))
```

#### Before
```python
assert result == {"columnname": "column=name"}
```

#### After
```python
assert result == {"columnname": "column+name", "c0_columnname": "column=name"}
```

File: src/datachain/lib/utils.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/lib/utils.py:44]: I'm fixing this (and a bunch of other case) - don't spend time yet on this please :)

File: src/datachain/lib/utils.py (hunk 3 above)
Comment (shcheklein) [reply src/datachain/lib/utils.py:54]: sorry about that, I'm fixing all of this right now ... give me a few minutes a review again pls

Comment (shcheklein) [reply src/datachain/lib/utils.py]: > I thought the whole point of this logic was checking for duplicates in the new_column_names dictionary?

it's not that simple ... we should also ideally check that there is no already a column in the original list and try to avoid naming it the same if possible

File: src/datachain/lib/utils.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/lib/utils.py]: not sure, I'll think a bit about this ... there could be non ascii literals - should we also change them to _ - not sure tbh.

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py]: You're right; we risk going too far with normalization, although I wonder if it'd be a good idea to create table names containing non-ASCII characters anyway.

We can always normalize diacritics with something like this:

```python
>>> from unicodedata import normalize
>>> normalize("NFKD", "próspera piraña").encode("ascii", "ignore")
"prospera pirana"
```

But, again, thorough normalization is a complex topic, and we'd still not be taking into account e.g. Cyrillic, Han...

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py]: Irrelevant rambling (continued)...

```suggestion
        new_column = org_column.casefold()  # Unicode-friendly .lower()
        new_column = regex.sub("[^\p{L}\p{Nd}]+", "_", new_column)  # Unicode-friendly [^a-z0-9]
        new_column = new_column.strip("_")
```

Requires the `regex` package to use Unicode character classes and crossing your fingers for database support of non-ASCII column names.

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py]: Based on the latest logic, I still believe my original proposal — https://github.com/iterative/datachain/pull/542#discussion_r1817963778 — would be an improvement over the current approach.

Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py]: Based on the latest logic, I still believe my original proposal — https://github.com/iterative/datachain/pull/542#discussion_r1817963778 — would be an improvement over the current approach.

File: src/datachain/lib/utils.py (hunk 1 above)
Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py]: Not sure if relevant, but in the same vein as column names beginning with a number aren't valid, reserved words like `return` shouldn't be valied either. 🙈

**Comment by 0x2b3bfa0** [comment]
I've been playing a bit with the logic and drafting an alternative implementation, with many debatable choices but some ideas that might be useful:

```python
def normalize_column_names(names: Sequence[str]) -> dict[str, str]:
    collision_counter = Counter(names)
    result = {}

    for original_name in names:
        normalized = original_name.lower()
        normalized = re.sub("[^0-9a-z]+", "_", normalized)
        normalized = normalized.strip("_")

        if not normalized:
            normalized = "_column"

        if not normalized.isidentifier() or iskeyword(normalized):
            normalized = f"_{normalized}"

        collision_counter.update({normalized: 1})
        if (collisions := collision_counter[normalized]) > 1:
            normalized = f"{normalized}_{collisions - 1}"

        result[normalized] = original_name

    return result
```

File: src/datachain/lib/utils.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/lib/utils.py]: changed to replace with `_` all "noisy" symbols, let's not go into things like unicode normalization, etc 

is there anything left here?

**Comment by shcheklein** [comment]
any other practical improvements left @0x2b3bfa0 ?

File: src/datachain/lib/utils.py [hunk 4]
```diff
@@ -28,3 +30,37 @@ def __init__(self, message):
...
+
+def normalize_col_names(col_names: Sequence[str]) -> dict[str, str]:
+    gen_col_counter = 0
+    new_col_names = {}
+    org_col_names = set(col_names)
+
+    for org_column in col_names:
+        new_column = org_column.lower()
+        new_column = re.sub("[^0-9a-z_\\s-]+", "_", new_column)
+        new_column = re.sub("[-_\\s]+", "_", new_column)
```
Comment (0x2b3bfa0) [inline src/datachain/lib/utils.py:43]: Can't we simplify all that regular expression soup like this?

```suggestion
        new_column = re.sub("[^0-9a-z]+", "_", new_column)
```

File: src/datachain/lib/utils.py [hunk 5]
```diff
@@ -28,3 +30,37 @@ def __init__(self, message):
...
+                if new_column:
+                    generated_column = f"{generated_column}_{new_column}"
+                if (
+                    generated_column not in org_col_names
+                    and generated_column not in new_col_names
+                ):
+                    new_column = generated_column
+                    break
+
+        new_col_names[new_column] = org_column
```
Comment (0x2b3bfa0) [inline src/datachain/lib/utils.py:64]:
```suggestion
        generated_column = new_column

        while (
            not generated_column
            or generated_column[0].isdigit()
            or generated_column in new_col_names
            or (generated_column != org_column and generated_column in org_col_names)
        ):
            if new_column:
                generated_column = f"c{gen_col_counter}_{new_column}"
            else:
                generated_column = f"c{gen_col_counter}"
            gen_col_counter += 1

        new_col_names[generated_column] = org_column
```

**Review by 0x2b3bfa0 — APPROVED** [review]
Made a couple PEP-20 suggestions, but looks good to me!

**Review by dreadatour — APPROVED** [review]
Looks like a good improvement to me 👍
I can’t think of any cases where this change would cause issues.

File: src/datachain/lib/utils.py (hunk 4 above)
Comment (0x2b3bfa0) [reply src/datachain/lib/utils.py:43]: I believe it's the same: we're replacing one or more non-alphanumerical characters with a single underscore, and it's a greedy match.

```python
>>> re.sub("[^0-9a-z]+", "_", "test$-_-$string")
'test_string'
```

**Comment by 0x2b3bfa0** [comment]
@shcheklein, are tests failing due to my last changes, or is it an unrelated failure?

**Comment by shcheklein** [comment]
> @shcheklein, are tests failing due to my last changes, or is it an unrelated failure?

no, it's an unfortunate coincidence - pyarrow 18 got released 8 hours ago

test is not reliable, I'll improve it a bit
